[Feature] 배틀 페이지 입장시 서버 데이터와 동기화한다

JuJangGwon opened the pull request:
## 🔗 관련 이슈
Closes #74

## ✅ 작업 내용
### 💡개선 사항
#### 1. 배틀 초기 진입 시 실시간 데이터 동기화
기존에는 UI 배치를 위해 임시용으로 하드코딩된 값을 사용했었는데, 현재 동기화 수신 이벤트도 구현됐기에 해당 이벤트를 통해 배틀 페이지 진입 시 최초 1회 실시간 데이터 동기화 하는 로직을 적용했습니다. **이로써 사용자가 게임 중간에 진입해도 문제 없이 게임 진행 할 수 있습니다!**

- **팀 인원수 동기화** - `battle:joined` 이벤트로 A/B팀 실시간 인원수 표시
- **Phase & Turn 상태 동기화** - 현재 진행 중인 게임 단계와 턴 상태를 정확히 반영
- **타임라인 데이터 동기화** - attack/defense 배열을 순차적으로 타임라인에 표기
- **투표 항목 동기화** - 턴 상태와 팀 조합으로 올바른 투표 리스트 매핑
- **채팅 라운지 인원 수 동기화** - 각 채팅 라운지에 현재 참여 중인 인원이 몇 명인지 표기 
- **채팅 내역 동기화** - 이 부분은 이미 #75 에서 구현됐기에 따로 건들진 않았습니다. 

<br/>

#### 2. 배틀 헤더에  방 제목과 방 설명 부분을 `/join` Api에서 가져온 실제 데이터로 표시하는 로직으로 변경
기존에 방 제목, 방 설명 text 내용을 모킹 용 상수 부분으로 채워져 있었는데,  `/join` Api에서 가져온 실제 데이터로 표시하여 동기화 시키는 로직을 적용 했습니다. 

<br/>

#### 3. 투표 시스템 아키텍처 임시 개선
현재 배틀페이지 index.tsx에 투표 관련 로직이 라인 분산되어 있어 중복 코드 발생 및 유지보수 어려움 겪고 있었습니다. 그래서 투표 상태 및 관련 처리 로직을 임시적으로 `useBattleSocket `훅으로 이동시켰습니다. (새롭게 추가될 투표 동기화 로직을 포함하면 거진 86줄..)

**개선 내용**:
- 투표 상태 관리 및 소켓 이벤트 처리 로직 `useBattleSocket`훅으로 이동 관련 사유는 고민 내용에 적어두겠습니다. 

<br/>

#### 4. API 엔드 포인트 URL 환경 변수 적용 
소켓 연결 URL이 하드코딩되어 있어 `개발/프로덕션` 환경 전환 시 코드를 수동으로 변경 해야 하는 상황이 생기는 것 같아서 환경 변수를 적용 했습니다. 

##### 기존 
```javascript
// useBattleSocket.ts
  // 배포 환경
    const newSocket = io("/", {
      transports: ['websocket']
    });
// dev 환경 아래처럼 수동 변경 필요
      const newSocket = io("http://localhost:3000", {
      transports: ['websocket']
    });
```
##### 이후
```javascript
  const newSocket = io(import.meta.env.VITE_API_URL, {
      transports: ['websocket']
    });
```
위 부분은 다음 스크럼에서 다시 한번 언급하겠습니다.  

<br/>

#### 5. 타입 시스템 개선
최초 동기화 할 때 받아오는 게임 데이터에 대한 타입을 선언 했었는데 (`BattleDiscussion`) 그 것을 깜빡하고 추가적으로  타임라인 컴포넌트 부분에서 `TimelineItemTypes`를 중복적으로 만들어 사용하고 있었던것을 확인했습니다. 그래서  `TimelineItemTypes`을 지워 불필요하게 중복 타입 사용하는 것을 개선했습니다. 

- `TimelineItemTypes` 제거 → `BattleDiscussion` 타입으로 통일
- 추가적으로 프론트 타입을 백엔드 응답 타입과 일치 작업 진행(누락된 필드를 추가했습니다.)

<br/>

#### 6. 서버에서 새로운 타임라인이 생길 시 전체 배열 저장에는 저장 안하는 버그 수정
##### 서버내의 배틀 상태 저장하는 객체 중 일부
```javascript
activeBattleState: {
  teamA: {
    ... //  <-  새로운 이의 제기/ 반론이 생겨도 추가 하고 있음
  },
  teamB: {
    ... //  <-  새로운 이의 제기/ 반론이 생겨도 추가 하고 있음
  },
  all: {
    ... //  <-  새로운 이의 제기/ 반론이 생겨도 추가 안하고 있었음.
  }
}
```
위의 간략한 예시처럼 새롭게 이의 제기/반론이 선정됐을 때  팀 배열에만 저장하고 전체 배열 저장 부분은 누락된 것으로 확인됐습니다. 그래서 전체 배열에도 선정된 이의 제기/반론 데이터를 저장하는 로직을 추가했습니다.

<br />

## 📸 참고 사항
### 배틀 페이지내 상태 / 비즈니스 로직의 분리가 매우 필요한 문제..
프로토타입 개발 마감이 다가올 수록 가독성 및 확장성 있는 코드를 고려하지 못하고 배틀 페이지 index.tsx에 다 때려박는 문제가 생겼던것 같습니다. 주로 소켓으로부터 게임 데이터 값을 받아와 변환, 처리하는 로직이 100줄이 넘을정도로 많았는데, 이거 때문에 투표 동기화 관련 로직을 추가 할 때도 복잡하고 난잡해서 시간이 더 걸렸던 것 같습니다. 이번 PR 작업하면서 이 부분을 같이 리펙토링 하고 싶었으나... 너무 방대한 작업으로 이어질 것 같아서 새롭게 이슈 만들어서 진행하겠습니다. 

방법은 고민중인데... 배틀 데이터 상태 처리 관련 내용을 `useBattleSocket.ts` 혼자 감당하기엔 너무 많습니다. 그리고 결합도도 너무 높고요.. 그래서 관심사에 따른 훅을 분리해서 만들어서 처리할 지 (ex. 투표 / 타임 라인 / 채팅 / 게임 진행) 아니면 zustand 나 reducer를 사용해서 store들을 분리 해볼 까 고민 중입니다. 

<br />

### `/joined` Event에서 중립 인원 수 필드 누락 문제 
<img width="122" height="78" alt="image" src="https://github.com/user-attachments/assets/3f5e1e38-747e-4881-9fa4-7f2980761c4b" />

게임 입장시 실시간 게임 참여 중인 진영 별 인원 수를 동기화 해야하는데, 반환 값에는 A팀, B팀 인원 수 밖에 없습니다.  해당 부분도 추후에 추가 해야할 것 같네요

JuJangGwon committed:
refactor : useBattleSocket에서 socket URL 하드 코딩으로 기입된 부분 환경 변수 적용 로직 적용
- 기존 '/' -> 이후 'VITE_API_URL'

JuJangGwon committed:
feat : 배틀 입장시 실시간 팀 인원수를 BattleHeader UI에 표시 하는 로직 적용
- 기존 하드 코딩된 값을 사용하고 있는 부분을 최초 소켓 입장시 서버와 동기화 용으로 받는 이벤트('battled:join') 반환 값을 사용하여 동기화

JuJangGwon committed:
fix : 타임라인 데이터를 전체 이의제기 배열에 저장안되는 버그 수정
- 원인 - 타임라인 데이터를 팀 배열에만 저장하고 전체 배열에 저장하는 로직이 누락 됐음
- activeBattleState.all에 저장

JuJangGwon committed:
refactor : 타임 라인 데이터 타입 통일 및 누락 필드 추가
- TimelineSection & TimeLineCard 에서 사용되는 TimelineItemTypes 제거 후 commons/types/battle.ts에 BattleDiscusstion Type으로 통일
- votes / status / authorId 필드 추가

JuJangGwon committed:
feat : 배틀페이지 최초 입장시 타임라인 데이터 동기화 하는 로직 적용
- battle/joined 이벤트 반환 값 중 timelines 데이터를 이용해 동기화 하는 로직 적용
- attack / defense 배열들에서 각각 한 개씩 순차적으로 타임라인에 표기하는 로직 적용

JuJangGwon committed:
feat : 배틀페이지 최초 입장시 게임 phaze & 턴 동기화 하는 로직 적용
- battle/joined 이벤트 반환 값 중 turn.status 데이터를 이용해 동기화 하는 로직 적용

JuJangGwon committed:
refactor : 투표 상태 및 투표 실시간 업데이트 로직을 useBattleSocket 훅으로 이동

- battlePage/index.tsx에서 관리하는 투표 상태 및 실시간 업데이트 로직을 hook에서 관리하도록 로직 수정
- 현재 상태 관련 로직이 battlePage/index.tsx에 많은 로직들이 집중되어있는데 점차 hook들로 분리 예정

JuJangGwon committed:
feat : 배틀페이지 최초 입장시 진행중인 투표 항목 동기화 로직 적용

- battle:joined 이벤트에서 현재 진행중인 투표 리스트 동기화 로직 적용
- 턴/팀 기반 투표 리스트 매핑 로직 구현

JuJangGwon committed:
fix : 새로고침 또는 재접속시 내 채팅이 채팅창 왼쪽에 표시되는 버그 수정
- 기존 하드 코딩으로 내 채팅인지 구분하는 로직에서 세션에 저장된 userId 기반으로 송신자 비교하는 로직으로 변경하여 새로고침 후에도 내가 쓴 채팅을 구분하는 로직 적용

JuJangGwon committed:
feat : 배틀 페이지 헤더에 방 제목 / 설명 UI에 /join APi로 부터 받은 값 반영하도록 적용
- 기존에는 방 제목/설명 부분이 하드코딩 되어있었어서, 배틀페이지 입장시 호출하는 /join APi 반환 값을 활용해 제대로 된 값을 반영 하도록 변경 했습니다.
- 이전에 /join APi 처리 로직을 적용할 때 적용한 기억이 있는데, 커밋 누락됐었던것 같습니다

JuJangGwon committed:
feat : 채팅 라운지 인원 수 표시 UI를 실제 데이터 기반으로 표시하도록 로직 변경

- 팀/전체 라운지 선택에 따른 인원 수를 동적 계산하는 로직 적용
- 임시용 하드코딩 부분 제거 후  최상단 부모에서 관리하는 현재 참가자 인원 관련 상태 데이터 기반으로 각각 채팅 라운지에 접속된 인원 수 표시 하도록 로직 수정

rhxfkg commented on frontend/src/pages/battlePage/components/timeline/TimelineSection.tsx:
```diff
@@ -1,56 +1,30 @@
 
-export default function TimelineSection() {
-  const [timeline] = useState<TimelineItem[]>(MOCK_TIMELINE);
+    for (let i = 0; i < maxLength; i++) {
+      if (i < attackList.length) {
+        result.push(attackList[i]);
+      }
+      if (i < defenseList.length) {
+        result.push(defenseList[i]);
+      }
+    }
```
수고하셨습니다!
attackList와 defenseList의 길이가 항상 같은 것이 가정인가요?
혹시 리스트에 추가되지 않는 경우가 있을 수도 있나요? (ex: 투표로 아무것도 올라오지 않았을 때)
만약 그런 경우가 있다면 오류가 발생할 수 있을 것 같습니당

HoonDongKang commented on frontend/src/pages/battlePage/hooks/useBattleSocket.ts:
```diff
@@ -45,8 +56,37 @@ export function useBattleSocket({ battleId, userId, team, password }: UseBattleS
+          'A_ATTACK-A': data.attacks,
+          'B_DEFENSE-B': data.defenses,
+          'B_ATTACK-B': data.attacks,
+          'A_DEFENSE-A': data.defenses
```
접미사?와 접두사로 들어가는 진영 이름이 동일한데 어떤 의미인가요??

HoonDongKang commented on frontend/src/pages/battlePage/hooks/useBattleSocket.ts:
```diff
@@ -95,11 +137,54 @@ export function useBattleSocket({ battleId, userId, team, password }: UseBattleS
+      });
+    };
+
+    newSocket.on('battle:attackvote:update', handleVoteUpdate);
+    newSocket.on('battle:defensevote:update', handleVoteUpdate);
+    newSocket.on('Battle:NewAttack', handleNewDiscussion);
+    newSocket.on('Battle:NewDefense', handleNewDiscussion);
```
컴포넌트 내부에서 `useEffect`로 리스닝을 해주면 반환할 때, `off`로 리스닝을 제거해주었던 거 같은데 훅으로 변경되면 그러한 로직은 사용하지 않아도 되나요??

JuJangGwon committed:
fix: 소켓 이벤트 리스너 메모리 누수 방지를 위한 cleanUp 추가

- useEffect cleanup에서 모든 socket.off() 호출 추가
- battle:closed, battle:chatUpdate 이벤트 cleanup 추가
- useBattleSocket의 11개 이벤트 리스너 cleanup 추가

JuJangGwon committed:
fix: 투표 매핑 로직 수정 및 상수 네이밍 컨벤션 적용

- VOTE_MAP 상수명을 UPPER_SNAKE_CASE로 변경
- 키 구분자를 언더스코어로 통일 (A_ATTACK-A → A_ATTACK_B)
- 투표 대상을 상대 팀으로 올바르게 매핑 하도록 로직 수정

JuJangGwon committed:
refactor: utils/getTurninfo.ts내 상수 네이밍 컨벤션을 UPPER_SNAKE_CASE로 통일

- phaseMap -> PHASE_MAP
- turnMap -> TURN_MAP

JuJangGwon commented:
## 리뷰 반영 사항

### 🐛 버그 수정

#### 1. 소켓 이벤트 리스너 메모리 누수 수정
- **문제**: useEffect cleanup에서 소켓 이벤트 리스너 제거하지 않아 컴포넌트 언마운트 시 메모리 누수 발생 가능 한 문제
- **해결**:
  - `index.tsx`: `battle:closed` 이벤트 cleanup 추가
  - `ChatSection.tsx`: `battle:chatUpdate` 이벤트 cleanup 추가  
  - `useBattleSocket.ts`: 11개 이벤트 리스너 개별 cleanup 추가

#### 2. 투표 매핑 로직 오류 수정
- **문제**: 투표 대상 팀 매핑이 잘못됨 (`A_ATTACK_A` → A팀이 A팀에게 투표)
- **해결**: 턴 주체의 상대 팀이 투표하도록 수정
  - `A_ATTACK_A` → `A_ATTACK_B` (A팀 공격 시 B팀이 투표)
  - `B_ATTACK_B` → `B_ATTACK_A` (B팀 공격 시 A팀이 투표)
  - 키 구분자 통일: 대시(`-`) → 언더스코어(`_`)

### ♻️ 리팩토링

#### 상수 네이밍 컨벤션 적용 (UPPER_SNAKE_CASE)
- `voteMap` → `VOTE_MAP` (useBattleSocket.ts)
- `phaseMap` → `PHASE_MAP` (getTurnInfo.ts)
- `turnMap` → `TURN_MAP` (getTurnInfo.ts)
